File: Sprint7/M/C++/main.cpp
// https://contest.yandex.ru/contest/25596/problems/M/
#include <iostream>
#include <vector>

int main() {
    int n, M;
    std::cin >> n >> M;

    std::vector<int> m(n + 1), c(n + 1);
    for(int i = 1; i <= n; ++i) {
        std::cin >> m[i] >> c[i];
    }

    std::vector<std::vector<int>> dp(n + 1, std::vector<int>(M + 1, 0));

    for(int i = 1; i <= n; ++i) {
        for(int j = 0; j <= M; ++j) {
            dp[i][j] = dp[i - 1][j];
            if(j >= m[i]) {
                dp[i][j] = std::max(dp[i][j], dp[i - 1][j - m[i]] + c[i]);
            }
        }
    }

    std::vector<int> result;
    int weight = M;

    for(int i = n; i >= 1; --i) {
        if(dp[i][weight] != dp[i-1][weight]) {
            result.push_back(i);
            weight -= m[i];
        }
    }

    std::cout << result.size() << std::endl;
    for(int i = result.size() - 1; i >= 0; --i) {
        std::cout << result[i] << " ";
    }

    return 0;
}

File: Sprint7/M/Rust/src/main.rs
// https://contest.yandex.ru/contest/25596/problems/M/

fn main() {
    let mut lines = std::io::stdin().lines();
    let line = lines.next().unwrap().unwrap();
    let mut n_m = line.split_whitespace();
    let n: usize = n_m.next().unwrap().parse::<usize>().unwrap();
    let m: usize = n_m.next().unwrap().parse::<usize>().unwrap();

    let mut weights = vec![0; n + 1];
    let mut values = vec![0; n + 1];

    for i in 1..=n {
        let line = lines.next().unwrap().unwrap();
        let mut parts = line.split_whitespace();
        weights[i] = parts.next().unwrap().parse::<usize>().unwrap();
        values[i] = parts.next().unwrap().parse::<usize>().unwrap();
    }

    let mut dp = vec![vec![0; m + 1]; n + 1];

    for i in 1..=n {
        for j in 0..=m {
            dp[i][j] = dp[i - 1][j];
            if j >= weights[i] {
                dp[i][j] = dp[i][j].max(dp[i - 1][j - weights[i]] + values[i]);
            }
        }
    }

    let mut result = Vec::new();
    let mut weight = m;

    for i in (1..=n).rev() {
        if dp[i][weight] != dp[i - 1][weight] {
            result.push(i);
            weight -= weights[i];
        }
    }

    println!("{}", result.len());
    for &item in result.iter().rev() {
        print!("{} ", item);
    }
}
